opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -1632,6 +1633,103 @@ namespace detail {
+        for (int w = 0; w < nw; ++w) {
+            const WellControls* ctrl = wells().ctrls[w];
+            const int nwc = well_controls_get_num(ctrl);
+            //Loop over all controls until we find a BHP control
+            //or a THP control that specifies what we need.
+            //Pick the value that gives largest potential flow
+            for (int ctrl_index=0; ctrl_index < nwc; ++ctrl_index) {
+
+                if (well_controls_iget_type(ctrl, ctrl_index) == BHP) {
+                    bhps[w] = well_controls_iget_target(ctrl, ctrl_index);
+                }
+
+                if(well_controls_iget_type(ctrl, ctrl_index) == THP) {
+                    double aqua = 0.0;
+                    double liquid = 0.0;
+                    double vapour = 0.0;
+
+                    if (active_[ Water ]) {
+                        aqua = well_state.wellRates()[w*np + pu.phase_pos[ Water ] ];
+                    }
+                    if (active_[ Oil ]) {
+                        liquid = well_state.wellRates()[w*np + pu.phase_pos[ Oil ] ];
+                    }
+                    if (active_[ Gas ]) {
+                        vapour = well_state.wellRates()[w*np + pu.phase_pos[ Gas ] ];
+                    }
+
+                    const int vfp        = well_controls_iget_vfp(ctrl, ctrl_index);
+                    const double& thp    = well_controls_iget_target(ctrl, ctrl_index);
+                    const double& alq    = well_controls_iget_alq(ctrl, ctrl_index);
+
+                    //Set *BHP* target by calculating bhp from THP
+                    const WellType& well_type = wells().type[w];
+
+                    const double gravity = detail::getGravity(geo_.gravity(), UgGridHelpers::dimensions(grid_));
+
+                    if (well_type == INJECTOR) {
+                        double dp = detail::computeHydrostaticCorrection(
+                                    wells(), w, vfp_properties_.getInj()->getTable(vfp)->getDatumDepth(),
+                                   stdWells().wellPerforationDensities(), gravity);
+                        const double bhp = vfp_properties_.getInj()->bhp(vfp, aqua, liquid, vapour, thp) - dp;
+                        // pick the bhp that gives the largest potentials i.e. largest bhp for injectors
+                        if ( bhp > bhps[w]) {
+                            bhps[w] = bhp;
+                        }
+                    }
+                    else if (well_type == PRODUCER) {
+                        double dp = detail::computeHydrostaticCorrection(
+                                    wells(), w, vfp_properties_.getProd()->getTable(vfp)->getDatumDepth(),
+                                    stdWells().wellPerforationDensities(), gravity);
+
+                        const double bhp = vfp_properties_.getProd()->bhp(vfp, aqua, liquid, vapour, thp, alq) - dp;
+                        // pick the bhp that gives the largest potentials i.e. smalest bhp for producers
```

Comment:
Is this correct? Should it not be the opposite (i.e. we must apply the strictest of the THP/BHP limits)?